fuel-p2p/src/orchestrator.rs

```diff
@@ -0,0 +1,98 @@
+use fuel_core_interfaces::model::Vote;
+use fuel_tx::Transaction;
+use libp2p::identity::Keypair;
+use tokio::sync::mpsc::{Receiver, Sender};
+
+use crate::{
+    behavior::FuelBehaviourEvent,
+    config::P2PConfig,
+    gossipsub::messages::GossipsubMessage,
+    service::{FuelP2PEvent, FuelP2PService},
+};
+
+pub struct NetworkOrchestrator {
+    p2p_service: FuelP2PService,
+
+    // receivers
+    rx_block_producer: Receiver<()>,
+    rx_bft: Receiver<Vote>,
+
+    // senders
+    tx_synchronizer: Sender<()>,
+    tx_transaction_pool: Sender<Transaction>,
+    tx_bft: Sender<()>,
+}
+
+impl NetworkOrchestrator {
+    pub async fn new(
+        local_keypair: Keypair,
+        p2p_config: P2PConfig,
+        rx_block_producer: Receiver<()>,
+        rx_bft: Receiver<Vote>,
+
+        tx_synchronizer: Sender<()>,
+        tx_transaction_pool: Sender<Transaction>,
+        tx_bft: Sender<()>,
+    ) -> Self {
+        let p2p_service = FuelP2PService::new(local_keypair, p2p_config)
+            .await
+            .unwrap();
+
+        Self {
+            p2p_service,
+
+            rx_bft,
+            rx_block_producer,
+
+            tx_bft,
+            tx_synchronizer,
+            tx_transaction_pool,
+        }
+    }
+
+    pub async fn run(&mut self) {
+        loop {
+            tokio::select! {
+                p2p_event = self.p2p_service.next_event() => {
+                    if let FuelP2PEvent::Behaviour(behaviour_event) = p2p_event {
+                        match behaviour_event {
+                            FuelBehaviourEvent::GossipsubMessage { message, topic_hash, peer_id } => {
```

Comment:
Gossipsub Messages, I would need to know for each message to which component I send it to.

Reply:
It can be modelled as broadcast, as you would have the ability to subscribe and sniff what is going on in the p2p layer without disturbing other parts of the system. https://github.com/FuelLabs/fuel-core/wiki/Interfaces-between-fuel-client-modules#p2p-channels
But it can be done with mpsc channel